crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs

```diff
@@ -66,57 +82,189 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
+            || (builtin_type != "float" && builtin_type != "int")
+        {
+            flat_nodes.push(expr);
         }
+    };
 
-        if has_int {
-            checker.diagnostics.push(Diagnostic::new(
-                RedundantNumericUnion {
-                    redundancy: Redundancy::IntComplex,
-                },
-                annotation.range(),
-            ));
+    // Traverse the union, and remember which numeric types are found.
+    traverse_union(&mut remove_numeric_type, checker.semantic(), annotation);
+
+    // Generate the flattened fix once.
+    let fix = if let &[fix_expr] = flat_nodes.as_slice() {
+        generate_single_fix(checker, fix_expr, annotation)
+    } else {
+        match union_type {
+            UnionLike::BinOp => generate_bit_or_fix(checker, flat_nodes, annotation),
+            UnionLike::TypingUnion => generate_union_fix(checker, flat_nodes, annotation)
+                .ok()
+                .unwrap(),
         }
-    } else if has_float && has_int {
-        checker.diagnostics.push(Diagnostic::new(
-            RedundantNumericUnion {
-                redundancy: Redundancy::IntFloat,
-            },
-            annotation.range(),
-        ));
+    };
+
+    checker.diagnostics.push(
+        Diagnostic::new(RedundantNumericUnion { redundancy }, annotation.range()).with_fix(fix),
+    );
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum Redundancy {
+    IntFloatComplex,
+    FloatComplex,
+    IntComplex,
+    IntFloat,
+}
+
+impl Redundancy {
+    pub(super) fn from_numeric_flags(numeric_flags: NumericFlags) -> Option<Self> {
+        match numeric_flags.bits() {
+            0b0110 => Some(Self::FloatComplex),
+            0b0111 => Some(Self::IntFloatComplex),
+            0b0101 => Some(Self::IntComplex),
+            0b0011 => Some(Self::IntFloat),
+            _ => None,
+        }
+    }
+}
+
+bitflags! {
+    #[derive(Copy, Clone, Debug, PartialEq, Eq)]
+    pub(super) struct NumericFlags: u8 {
+        /// `int`
+        const INT = 0b0001;
+        /// `float`
+        const FLOAT = 0b0010;
+        /// `complex`
+        const COMPLEX = 0b0100;
     }
 }
+
+impl NumericFlags {
+    pub(super) fn seen_builtin_type(&mut self, name: &str) {
+        let flag: NumericFlags = match name {
+            "int" => NumericFlags::INT,
+            "float" => NumericFlags::FLOAT,
+            "complex" => NumericFlags::COMPLEX,
+            _ => {
+                return;
+            }
+        };
+        self.insert(flag);
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum UnionLike {
+    /// E.g., `typing.Union[int, str]`
+    TypingUnion,
+    /// E.g., `int | str`
+    BinOp,
+}
+
+// Generate a [`Fix`] for two or more type expressions, e.g. `int | float | complex`.
+fn generate_bit_or_fix(checker: &Checker, nodes: Vec<&Expr>, annotation: &Expr) -> Fix {
+    debug_assert!(nodes.len() >= 2, "At least two nodes required");
```

Comment:
The code above handles `nodes.len() < 2` separately and thus this function should never encounter these cases. Is `debug_assert!` ok to denote this?